core/src/main/java/org/apache/iceberg/stats/StatsReader.java

```diff
@@ -0,0 +1,183 @@
+      Preconditions.checkState(
+          fileFormatVersion == StatsFormat.CURRENT_FORMAT_VERSION,
+          "Unsupported format version %s, %s is the latest supported",
+          fileFormatVersion,
+          StatsFormat.CURRENT_FORMAT_VERSION);
+
+      int flags = StatsFormat.readIntegerLittleEndian(footer, footerSize - 12);
+      Preconditions.checkState((flags & ~StatsFormat.SUPPORTED_FLAGS) == 0, "Unsupported flags: %s", flags);
+      boolean compressed = isFlagSet(flags, StatsFormat.FLAG_COMPRESSED);
+
+      int reserved = StatsFormat.readIntegerLittleEndian(footer, footerSize - 16);
+      Preconditions.checkState(reserved == 0, "Unexpected reserved bytes value: %s", reserved);
+
+      int footerPayloadSize = StatsFormat.readIntegerLittleEndian(footer, footerSize - 20);
+      Preconditions.checkState(footerPayloadSize == footerSize - 24,
+          "Unexpected footer payload size value %s for footer size %s", footerPayloadSize, footerSize);
+
+      FileMetadata readFileMetadata;
+      ByteBuffer footerPayload = ByteBuffer.wrap(footer, 4, footerPayloadSize);
+      if (!compressed) {
+        readFileMetadata = parseFileMetadata(footerPayload);
+      } else {
+        ByteBuffer footerJson = StatsFormat.decompressFooterPayload(footerPayload);
+        readFileMetadata = parseFileMetadata(footerJson);
+      }
+      this.knownFileMetadata = Optional.of(readFileMetadata);
+    }
+    return knownFileMetadata.get();
+  }
+
+  /**
+   * @throws IOException          when I/O error occurs
+   * @throws UncheckedIOException when I/O error occurs
+   */
+  // Note: The method is marked as throwing IOException to allow future implementation evolution.
+  // Currently, the exception isn't being thrown.
+  public Stream<Map.Entry<BlobMetadata, ByteBuffer>> readAll(List<BlobMetadata> blobs) throws IOException {
+    if (blobs.isEmpty()) {
+      return Stream.empty();
+    }
+
+    // TODO inspect blob offsets and coalesce read regions close to each other
+
+    return blobs.stream()
+        .sorted(Comparator.comparingLong(BlobMetadata::offset))
+        .map((BlobMetadata blobMetadata) -> {
+          try {
+            input.seek(blobMetadata.offset());
+            byte[] bytes = new byte[Math.toIntExact(blobMetadata.length())];
+            ByteStreams.readFully(input, bytes);
+            ByteBuffer data = ByteBuffer.wrap(bytes);
+            if (blobMetadata.compressionCodec() != null) {
+              StatsCompressionCodec codec = StatsCompressionCodec.forName(blobMetadata.compressionCodec());
+              data = StatsFormat.decompressBlob(codec, data);
+            }
+            return new SimpleEntry<>(blobMetadata, data);
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+  }
+
+  private static void checkMagic(byte[] data, int offset) {
+    int read = StatsFormat.readIntegerLittleEndian(data, offset);
```

Comment:
I think converting the magic number to an int makes this code harder to reason about. Magic is not checked in a tight loop, so I don't think that the optimization of comparing the bytes all at once is worth it. I'd prefer to just check each byte in a loop and avoid needing to validate that this is correct:

```java
static final int MAGIC_AS_NUMBER_LE = new BigInteger(swapBytes(getMagic())).intValueExact();
```